### PR TITLE
gh-102821: Simplify `threading.Event.wait` logic

### PR DESCRIPTION
**Pull request by mementum**
https://github.com/python/cpython/issues/102821

* Issue: gh-102821

This is the same as #102822 but without the PEP-8 text changes, just the logic of "threading.Event.wait"

### CODE REVIEW

**Comment by JosephSBoyle** [comment]
Hi @mementum, I think you need to prefix the name of this PR with `gh-102821` in order for the bots to setup a link between the issue and PR correctly:)

**Comment by mementum** [comment]
> Hi @mementum, I think you need to prefix the name of this PR with `gh-102821` in order for the bots to setup a link between the issue and PR correctly:)

Indeed. I did that with care for the 1st pull-request for the issue and was careless this time.

Thanks @AlexWaygood for having done the edit.

**Comment by Eclips4** [comment]
Looks like you did what needed to be done =)
Tests are passed, so we can say this changes really simplified logic.
It remains to wait, what core developers will say.

Also, @AlexWaygood, can u set label `skip news`? There's only "cosmetical" changes, so, I think, there is no need in NEWS entry.

**Comment by AlexWaygood** [comment]
> Also, @AlexWaygood, can u set label `skip news`? There's only "cosmetical" changes, so, I think, there is no need in NEWS entry.

No, I'm not going to add the "skip news" label. I haven't looked at this PR yet, but in general we reject PRs in the CPython repo if they are purely cosmetic and have no user-facing impact. The only reason we might accept this PR is if it led to a meaningful performance improvement -- and if it leads to a meaningful performance improvement, then it deserves a news entry.

**Comment by Eclips4** [comment]
@mementum, as Alex said, this PR would be merged if it led to a meaningful performance improvement.
For this, you can use such tool as `pyperformance`, after, please, share results. Good luck!

**Comment by AlexWaygood** [comment]
> For this, you can use such tool as `pyperformance`

A microbenchmark would be fine for this PR, since it's a small PR affecting an isolated piece of logic :)

**Comment by Eclips4** [comment]
> > For this, you can use such tool as `pyperformance`
> 
> A microbenchmark using `timeit` would be fine for this PR, since it's a small PR affecting an isolated piece of logic :)

As I understand, this only can affect on `Thread.start` because it called `self._started.wait()` (`self._started` is  represented by instance of class Event). In the rest of module i cannot find usages of this. So you're right Alex, `timeit` with code which length about ~ten lines will be enough.

**Comment by JosephSBoyle** [comment]
On my desktop Python3.10.1 on Win64
```
OtherEvent: [0.3933764999965206, 0.3920841999934055, 0.39645719999680296, 0.39919389999704435, 0.3962173000036273]
Event: [0.4053313000185881, 0.4035691999888513, 0.4008484999940265, 0.40631520000169985, 0.4063303000002634]      
% improvement: [2.949389800767732, 2.8458564220865012, 1.095501167470747, 1.7526540982532035, 2.4888618930533846]
```

**Comment by mementum** [comment]
On a Linux machine: Ubuntu 22.04 - Python 3.9.13 

```bash
(base) occ@CO-03884:~/tmp$ uname -a
Linux CO-03884 5.15.0-67-generic #74-Ubuntu SMP Wed Feb 22 14:14:39 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```

```bash
(base) occ@CO-03884:~/tmp$ python event-wait-logic-test.py
OtherEvent: [0.5852362000150606, 0.5834260219708085, 0.583657459996175, 0.5832577000255696, 0.5842611940461211]
Event: [0.5966716720140539, 0.5977747769793496, 0.603394869947806, 0.5979562159627676, 0.59788854297949]
% improved: [1.916543475307464, 2.400361400500628, 3.271060284841054, 2.458125786606624, 2.2792457044684356]
```

File: Misc/NEWS.d/next/Library/2023-03-22-19-41-23.gh-issue-102821.nAt1VY.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Simplify redundant logic of :meth:`threading.Event.wait`.
```
Comment (AlexWaygood) [inline Misc/NEWS.d/next/Library/2023-03-22-19-41-23.gh-issue-102821.nAt1VY.rst:1]: This news entry isn't informative about the impact this patch will have on users of `threading` (the performance boost). It only tells me about the changes to the internal implementation, which users of the `threading` module likely won't care about.

**Comment by AlexWaygood** [comment]
> The microbenchmark (Edit: Windows 11 22H2 Build 22621.1413 - Python 3.10.9)

A benchmark using Python 3.10 is meaningless, since significant changes have been made to the interpreter since 3.10 which means that many low-level things (which may include assignments of local variables) are much faster on the `main` branch than they were in 3.10.

Using a simpler benchmark, and running it on `main`, I cannot see any meaningful performance improvement from this PR:

```
C:\Users\alexw\coding\cpython>gh pr checkout https://github.com/python/cpython/pull/102843
remote: Enumerating objects: 14, done.
remote: Counting objects: 100% (10/10), done.
remote: Total 14 (delta 10), reused 10 (delta 10), pack-reused 4
Unpacking objects: 100% (14/14), 13.12 KiB | 151.00 KiB/s, done.
From https://github.com/python/cpython
 * [new ref]               refs/pull/102843/head -> event-wait-logic-simplify
Switched to branch 'event-wait-logic-simplify'

C:\Users\alexw\coding\cpython>python -m timeit -s "import threading; e = threading.Event(); e.set()" "e.wait()"
Running PGUpdate|x64 interpreter...
1000000 loops, best of 5: 369 nsec per loop

C:\Users\alexw\coding\cpython>python -m timeit -s "import threading; e = threading.Event(); e.set()" "e.wait()"
Running PGUpdate|x64 interpreter...
1000000 loops, best of 5: 367 nsec per loop

C:\Users\alexw\coding\cpython>python -m timeit -s "import threading; e = threading.Event(); e.set()" "e.wait()"
Running PGUpdate|x64 interpreter...
1000000 loops, best of 5: 372 nsec per loop

C:\Users\alexw\coding\cpython>git checkout main
Switched to branch 'main'
Your branch is up to date with 'origin/main'.

C:\Users\alexw\coding\cpython>python -m timeit -s "import threading; e = threading.Event(); e.set()" "e.wait()"
Running PGUpdate|x64 interpreter...
1000000 loops, best of 5: 373 nsec per loop

C:\Users\alexw\coding\cpython>python -m timeit -s "import threading; e = threading.Event(); e.set()" "e.wait()"
Running PGUpdate|x64 interpreter...
1000000 loops, best of 5: 378 nsec per loop

C:\Users\alexw\coding\cpython>python -m timeit -s "import threading; e = threading.Event(); e.set()" "e.wait()"
Running PGUpdate|x64 interpreter...
1000000 loops, best of 5: 380 nsec per loop
```

I also doubt that calling `threading.Event.wait` is ever likely to be performance-critical. As such, I'm closing this PR.

**Comment by AlexWaygood** [comment]
@mementum, I'd encourage you to look at our [6,800 open issues](https://github.com/python/cpython/issues) if you're interested in contributing to CPython :)

**Comment by mementum** [comment]
I had a real feature proposal at hand

See: https://github.com/python/cpython/issues/102983

(In any case, I would be interested to learn if there is any thread for meta-discussions on the simplification/modernization of modules in the `stdlib`)

**Comment by AlexWaygood** [comment]
> I had a real feature proposal at hand

Well, that wasn't mentioned in either your issue or your PR description. I was asked to consider this PR in isolation, so I considered it in isolation.

> (In any case, I would be interested to learn if there is any thread for meta-discussions on the simplification/modernization of modules in the `stdlib`)

As I explained above: in general, we don't encourage PRs simplifying or modernising stdlib code, as there's too much risk of inadvertently breaking somebody else's code, even with a seemingly innocuous change. The stdlib needs to be extremely stable. We also have many, many open bug reports and prefer to spend our time trying to fix those rather than reviewing PRs attempting to modernise code that has no known bugs in it.

**Comment by mementum** [comment]
Not a related "feature". It was just  a comment. I actually came to this small (now rejected) proposal for `threading.Event` because I wanted to see the details in the `threading` module to consider how to best implement my idea for `queue.Queue`

**Pull request closed without merging**